File: frontend/src/app/core/services/gemini.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';

/**
 * Estructura esperada del JSON que devuelve Gemini al analizar una planilla.
 */
export interface OcrAttendanceResult {
  /** Mes de la planilla (1-12 o texto como "Abril") */
  month?: string | number;
  /** Año de la planilla */
  year?: number;
  /** Nombre del vigilador encontrado en la imagen */
  guardName?: string;
  /** Nombre del objetivo/lugar de trabajo */
  workplace?: string;
  /** Filas diarias extraídas */
  rows: OcrRow[];
}

export interface OcrRow {
  /** Número de día (1-31) */
  day: number;
  /** Hora de entrada en formato HH:mm o vacío */
  entry: string;
  /** Hora de salida en formato HH:mm o vacío */
  exit: string;
  /** ¿Es franco/día libre? */
  isDayOff: boolean;
  /** Observaciones opcionales */
  notes: string;
}

const OCR_PROMPT = `Analizá esta planilla de control de horarios de vigilancia.
Extraé los datos de cabecera y el detalle de cada día.

Devolvé ESTRICTAMENTE solo un objeto JSON válido (sin markdown, sin bloques de código, sin texto adicional) con esta estructura exacta:
{
  "month": "<mes como número 1-12 o nombre en español>",
  "year": <año como número>,
  "guardName": "<apellido y nombre del vigilador si se ve>",
  "workplace": "<nombre del objetivo o lugar si se ve>",
  "rows": [
    {
      "day": <número de día>,
      "entry": "<hora entrada en formato HH:mm, o vacío si no hay>",
      "exit": "<hora salida en formato HH:mm, o vacío si no hay>",
      "isDayOff": <true si es franco/descanso, false si no>,
      "notes": "<observaciones si las hay, o cadena vacía>"
    }
  ]
}

Si un campo no se puede leer con claridad, dejalo como vacío ("") o null.
Los horarios deben estar en formato HH:mm de 24 horas.
No incluyas días que estén completamente en blanco si el mes tiene menos de 31 días.`;

const SYSTEM_INSTRUCTION = 'Sos un sistema experto en lectura óptica de planillas de control de asistencia y horarios. Tu única tarea es extraer datos y devolver JSON estructurado.';

@Injectable({
  providedIn: 'root'
})
export class GeminiService {

  private readonly apiUrl = 'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-lite:generateContent';

  /**
   * Procesa un archivo de imagen y extrae los datos de la planilla usando Gemini Vision.
   * @param imageFile Archivo de imagen de la planilla manuscrita
   * @returns Promise con los datos extraídos
   */
  async processAttendanceSheet(imageFile: File): Promise<OcrAttendanceResult> {
    const base64 = await this.fileToBase64(imageFile);

    const fullPrompt = `${SYSTEM_INSTRUCTION}\n\n${OCR_PROMPT}`;

    const body = {
      contents: [
        {
          role: 'user',
          parts: [
            { text: fullPrompt },
            {
              inlineData: {
                mimeType: imageFile.type,
                data: base64
              }
            }
          ]
        }
      ]
    };

    const url = `${this.apiUrl}?key=${environment.geminiApiKey}`;

    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body)
    });

    if (!response.ok) {
      const errBody = await response.text();
      let detail = `Error ${response.status}`;
      try {
        const parsed = JSON.parse(errBody);
        detail = parsed?.error?.message || detail;
      } catch { /* ignore */ }
      throw new Error(`Gemini API respondió con error: ${detail}`);
    }

    const data = await response.json();
    const text = data?.candidates?.[0]?.content?.parts?.[0]?.text?.trim();

    if (!text) {
      throw new Error('Gemini no devolvió contenido de texto en la respuesta.');
    }

    // Limpiar posibles bloques markdown que el modelo agregue
    const jsonText = text
      .replace(/^```json\s*/i, '')
      .replace(/^```\s*/i, '')
      .replace(/\s*```$/i, '')
      .trim();

    try {
      return JSON.parse(jsonText) as OcrAttendanceResult;
    } catch {
      throw new Error(`La respuesta de la IA no es un JSON válido. Respuesta recibida:\n${text.substring(0, 300)}`);
    }
  }

  private fileToBase64(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => {
        const dataUrl = reader.result as string;
        // Extraer solo el contenido base64 sin el prefijo "data:image/...;base64,"
        resolve(dataUrl.split(',')[1]);
      };
      reader.onerror = () => reject(new Error('No se pudo leer el archivo de imagen.'));
      reader.readAsDataURL(file);
    });
  }
}

File: backend/SGV.Business/Services/GeminiService.cs
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Configuration;
using SGV.Business.Interfaces;
using SGV.DTOs.Ocr;

namespace SGV.Business.Services
{
    public class GeminiService : IGeminiService
    {
        private readonly HttpClient _httpClient;
        private readonly string? _apiKey;
        private const string ApiUrl = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-lite:generateContent";
        private const string SystemInstruction = "Sos un sistema experto en lectura óptica de planillas de control de asistencia y horarios. Tu única tarea es extraer datos y devolver JSON estructurado.";
        private const string OcrPrompt = @"Analizá esta planilla de control de horarios de vigilancia.
Extraé los datos de cabecera y el detalle de cada día.

Devolvé ESTRICTAMENTE solo un objeto JSON válido (sin markdown, sin bloques de código, sin texto adicional) con esta estructura exacta:
{
  ""month"": ""<mes como número 1-12 o nombre en español>"",
  ""year"": <año como número>,
  ""guardName"": ""<apellido y nombre del vigilador si se ve>"",
  ""workplace"": ""<nombre del objetivo o lugar si se ve>"",
  ""rows"": [
    {
      ""day"": <número de día>,
      ""entry"": ""<hora entrada en formato HH:mm, o vacío si no hay>"",
      ""exit"": ""<hora salida en formato HH:mm, o vacío si no hay>"",
      ""isDayOff"": <true si es franco/descanso, false si no>,
      ""notes"": ""<observaciones si las hay, o cadena vacía>""
    }
  ]
}

Si un campo no se puede leer con claridad, dejalo como vacío ("""") o null.
Los horarios deben estar en formato HH:mm de 24 horas.
No incluyas días que estén completamente en blanco si el mes tiene menos de 31 días.";

        public GeminiService(HttpClient httpClient, IConfiguration configuration)
        {
            _httpClient = httpClient;
            _apiKey = configuration["GeminiApiKey"];
        }

        public async Task<OcrAttendanceResultDto> ProcessAttendanceSheetAsync(Stream imageStream, string contentType)
        {
            if (string.IsNullOrEmpty(_apiKey))
                throw new Exception("Gemini API Key is not configured in the backend (GeminiApiKey).");

            string base64Image;
            using (var ms = new MemoryStream())
            {
                await imageStream.CopyToAsync(ms);
                base64Image = Convert.ToBase64String(ms.ToArray());
            }

            var requestBody = new
            {
                contents = new[]
                {
                    new
                    {
                        role = "user",
                        parts = new object[]
                        {
                            new { text = $"{SystemInstruction}\n\n{OcrPrompt}" },
                            new
                            {
                                inlineData = new
                                {
                                    mimeType = contentType,
                                    data = base64Image
                                }
                            }
                        }
                    }
                }
            };

            var jsonBody = JsonSerializer.Serialize(requestBody);
            var content = new StringContent(jsonBody, Encoding.UTF8, "application/json");

            var url = $"{ApiUrl}?key={_apiKey}";
            var response = await _httpClient.PostAsync(url, content);

            if (!response.IsSuccessStatusCode)
            {
                var errorBody = await response.Content.ReadAsStringAsync();
                throw new Exception($"Gemini API error: {response.StatusCode}. Details: {errorBody}");
            }

            var responseBody = await response.Content.ReadAsStringAsync();
            var jsonDocument = JsonDocument.Parse(responseBody);
            
            var root = jsonDocument.RootElement;
            if (root.TryGetProperty("candidates", out var candidates) && candidates.GetArrayLength() > 0)
            {
                var firstCandidate = candidates[0];
                if (firstCandidate.TryGetProperty("content", out var contentElement) && 
                    contentElement.TryGetProperty("parts", out var parts) && 
                    parts.GetArrayLength() > 0)
                {
                    var text = parts[0].GetProperty("text").GetString()?.Trim();
                    
                    if (string.IsNullOrEmpty(text))
                        throw new Exception("Gemini returned empty text.");

                    // Clean markdown blocks if any
                    if (text.StartsWith("```json", StringComparison.OrdinalIgnoreCase))
                        text = text.Substring(7);
                    else if (text.StartsWith("```"))
                        text = text.Substring(3);

                    if (text.EndsWith("```"))
                        text = text.Substring(0, text.Length - 3);

                    text = text.Trim();

                    try
                    {
                        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
                        var result = JsonSerializer.Deserialize<OcrAttendanceResultDto>(text, options);
                        if (result != null)
                            return result;
                    }
                    catch (JsonException ex)
                    {
                        throw new Exception($"Failed to parse JSON from Gemini. Exception: {ex.Message}. Response text: {text.Substring(0, Math.Min(300, text.Length))}");
                    }
                }
            }

            throw new Exception("Could not parse response structure from Gemini API.");
        }
    }
}
